src/dmd/pragmasem.d

```diff
@@ -0,0 +1,489 @@
+{
+    Scope* sc;
+    Expressions* args;
+    PragmaStatement ps;
+    PragmaDeclaration pd;
+    Statement result;
+
+    this(Scope* sc, PragmaStatement ps, PragmaDeclaration pd)
+    {
+        this.sc = sc;
+        this.args = ps ? ps.args : pd.args;
+        this.ps = ps;
+        this.pd = pd;
+    }
+
+    private void setError()
+    {
+        if (ps)
+            result = new ErrorStatement();
+    }
+
+    // Because auto p = ps ? ps : pd; p.error("foo"); does not compile
+    private void error(const(char)* msg, const(char)* msg2 = null)
+    {
+        if (msg2)
+        {
+            if (ps)
+                ps.error(msg, msg2);
+            else
+                pd.error(msg, msg2);
+        }
+        else
+        {
+            if (ps)
+                ps.error(msg);
+            else
+                pd.error(msg);
+        }
+        setError();
+    }
+
+    void declSementic()
+    {
+        if (pd && pd.decl)
+        {
+            pd.error("is missing a terminating `;`");
+            Scope* sc2 = pd.newScope(sc);
+            foreach(s; *pd.decl)
+            s.dsymbolSemantic(sc2);
+            
+            if (sc2 != sc)
+            sc2.pop();
+        }
+    }
+    void semantic(Identifier ident)
+    {
+        import dmd.target;
+        
+        if (ident == Id.msg)
+            pragmaMsgSemantic();
+        else if (ident == Id.lib)
+            pragmaLibSemantic();
+        else if (ident == Id.startaddress)
+            pragmaStartAddressSemantic();
+        else if (ident == Id.Pinline)
+            pragmaInlineSemantic();
+        else if (ident == Id.mangle)
+            pragmaMangleSemantic(); // does semantic on its decls
+        else if (ident == Id.crt_constructor || ident == Id.crt_destructor)
+            pragmaCrtCtorDtorSemantic(ident);
+        else if (global.params.ignoreUnsupportedPragmas && global.params.verbose)
+        {
+            /* Print unrecognized pragmas
+             */
+            OutBuffer buf;
+            buf.writestring(pd.ident.toChars());
+            bool first = true;
+            foreach (e; *args)
+            {
+                sc = sc.startCTFE();
+                e = e.expressionSemantic(sc);
+                e = resolveProperties(sc, e);
+                sc = sc.endCTFE();
+                e = e.ctfeInterpret();
+                if (first)
+                {
+                    buf.writestring(" (");
+                    first = false;
+                }
+                else
+                buf.writeByte(',');
+                buf.writestring(e.toChars());
+            }
+            if (args.dim)
+            buf.writeByte(')');
+            message("pragma    %s", buf.peekString());
+        }
+        else
+        {
+            enum msg = "unrecognized `pragma(%s)`";
+            if (ps)
+                .error(ps.loc, msg, ps.ident.toChars());
+            else
+                .error(pd.loc, msg, pd.ident.toChars());
```

Comment:
Not sure why a global `.error` call, given that above there's a custom `error`.

Reply:
Good point, I think I was confused about the `loc` parameter. It should fit the custom one.